Update useful links cypress tests

Bopchy opened the pull request:
Resolves #4496 

**Overall change:** _Update useful links cypress test so that it checks for presence of a useful link._

**Code changes:**

- _Add `data-e2e` attribute to useful link._
- _Update snapshots._
- _Add e2e test that checks for useful link items._
---

- [x] I have assigned myself to this PR and the corresponding issues
- [x] Automated (jest and/or cypress) tests added (for new features) or updated (for existing features)
- [ ] If necessary, I have run the local E2E non-smoke tests relevant to my changes (`CYPRESS_APP_ENV=local CYPRESS_SMOKE=false npm run test:e2e:interactive`) 
- [x] This PR requires manual testing

Commit:
Add data attribute. Add useful link test.

Commit:
Update snapshots.

Commit:
Add data-e2e="useful-link-item" attribute. Check for nested useful links.

Commit:
Update snapshots.

Commit:
Create shouldDisplayUsefulLinks().

thekp committed:
Merge branch 'latest' into 4496-update-useful-links-cypress-tests

OlgaLyubin commented on cypress/integration/pages/frontPage/tests.js:
```diff
@@ -6,6 +6,32 @@ const serviceHasIndexAlsos = service => service === 'thai';
+  if (usefulLinks) {
+    usefulLinksItems = usefulLinks.items.filter(item => {
+      return (
+        item.assetTypeCode === 'PRO' && contentTypes.includes(item.contentType)
```
Should content type not be 'Guide' for useful links? Or was there a decision to support all of the types from the `contentTypes` array above?

Bopchy replied:
Yes, for useful links the contentType should be `Guide`. Changed the function to only filter `item.contentType === 'Guide'`

OlgaLyubin committed:
Merge branch 'latest' into 4496-update-useful-links-cypress-tests

HarveyPeachey commented on cypress/integration/pages/frontPage/tests.js:
```diff
@@ -168,42 +194,25 @@ export const testsThatFollowSmokeTestConfig = ({ service, pageType }) =>
-                //  We include the hasStrapline as we don't render Useful Links
-                //  if we don't receive a strapline in the data
-                cy.get('[aria-labelledby="Useful-links"]').should('be.visible');
+                cy.get('div[data-e2e=useful-links]')
```
Do we need to check for `data-e2e=useful-links`? Can't we just check for the `data-e2e="useful-link-item"` if the page should display useful links?

Bopchy replied:
Yes, we can just check for the `useful-link-item`. Updated test to check for the item only

DenisHdz replied:
With this, we will be rendering a data attribute for each useful link item and I am not sure if it is the best approach. I would rather add it just to the section and then look within it that there are link items, as we do with the `Index Alsos`. Happy to hear other opinions on this.

Bopchy replied:
Opted to keep `data-e2e="useful-link-item"` as I figured it would be an easy approach to check the items. Understand the concern though. Changed to using an attribute on the wrapper and looking within it for useful link `<a>'s`

sadickisaac commented on cypress/integration/pages/frontPage/tests.js:
```diff
@@ -6,6 +6,32 @@ const serviceHasIndexAlsos = service => service === 'thai';
 const serviceHasPublishedPromo = service => service === 'persian';
 
+// Check for valid useful links
+const shouldDisplayUsefulLinks = pageData => {
```
What do you think about renaming this to something more clear. i.e `isValidUsefulLinks`. This communicates what the function is doing. Checking for `usefulLinks` validity.
```js
/* This gives an impression that we are displaying usefulLinks inside the if block. */
if(shouldDisplayUsefulLinks) {
}

/* This gives an impression that we want to perform some action based on the validity of usefulLinks */
if(isValidUsefulLinks) {
}
```

sadickisaac commented on cypress/integration/pages/frontPage/tests.js:
```diff
@@ -168,42 +194,25 @@ export const testsThatFollowSmokeTestConfig = ({ service, pageType }) =>
+              if (shouldDisplayUsefulLinks(pageData)) {
+                // We include the hasStrapline in shouldDisplayUsefulLinks
+                // as we don't render Useful Links if we don't receive a
+                // strapline in the data
```
Should this comment be in the `shouldDisplayUsefulLinks` function where we check for validity of the `usefulLinks`?

Bopchy replied:
Yes. Moved it to the now  `isValidUsefulLinks()`

Commit:
Merge branch 'latest' into 4496-update-useful-links-cypress-tests

Commit:
Make changes
- Change shouldDisplayUsefulLinks() to isValidUsefulLinks().
- Filter against `item.contentType === 'Guide'` and not all contentTypes.
- Remove `data-e2e="useful-link"` visibility check

Commit:
Remove `data-e2e="useful-link(s)"` attributes. Update snapshot.

Commit:
Merge branch 'latest' into 4496-update-useful-links-cypress-tests

HarveyPeachey reviewed (approved):
LGTM 👍

OlgaLyubin committed:
Merge branch 'latest' into 4496-update-useful-links-cypress-tests

EinsteinNjoroge committed:
Merge branch 'latest' into 4496-update-useful-links-cypress-tests

EinsteinNjoroge commented on cypress/integration/pages/frontPage/tests.js:
```diff
@@ -6,6 +6,24 @@ const serviceHasIndexAlsos = service => service === 'thai';
+    return data.type === 'useful-links';
+  });
+
+  if (usefulLinks) {
```
```suggestion
  if (usefulLinks && 'strapline' in usefulLinks) {
```

EinsteinNjoroge commented on cypress/integration/pages/frontPage/tests.js:
```diff
@@ -6,6 +6,24 @@ const serviceHasIndexAlsos = service => service === 'thai';
+
+  // We include a check for the strapline as we don't render Useful Links
+  // if we don't receive a strapline in the data
+  return usefulLinks && 'strapline' in usefulLinks && usefulLinksItems.length;
```
```suggestion
  return false;
```

EinsteinNjoroge commented on cypress/integration/pages/frontPage/tests.js:
```diff
@@ -6,6 +6,24 @@ const serviceHasIndexAlsos = service => service === 'thai';
+  });
+
+  if (usefulLinks) {
+    usefulLinksItems = usefulLinks.items.filter(item => {
```
```suggestion
    return usefulLinks.items.some(
    ({ assetTypeCode, contentType }) =>
      assetTypeCode === 'PRO' && contentType === 'Guide',
  );
```

EinsteinNjoroge commented on cypress/integration/pages/frontPage/tests.js:
```diff
@@ -6,6 +6,24 @@ const serviceHasIndexAlsos = service => service === 'thai';
+
+  if (usefulLinks) {
+    usefulLinksItems = usefulLinks.items.filter(item => {
+      return item.assetTypeCode === 'PRO' && item.contentType === 'Guide';
```
```suggestion
```

EinsteinNjoroge commented on cypress/integration/pages/frontPage/tests.js:
```diff
@@ -6,6 +6,24 @@ const serviceHasIndexAlsos = service => service === 'thai';
+  if (usefulLinks) {
+    usefulLinksItems = usefulLinks.items.filter(item => {
+      return item.assetTypeCode === 'PRO' && item.contentType === 'Guide';
+    });
```
```suggestion
```

EinsteinNjoroge reviewed (commented):
Consider refactoring

DenisHdz reviewed (changes requested):
https://github.com/bbc/simorgh/pull/4587#discussion_r345640898
https://github.com/bbc/simorgh/pull/4587#pullrequestreview-316215388

Commit:
Add wrapper attribute.

Commit:
Merge branch 'latest' of github.com:bbc/simorgh into 4496-update-useful-links-cypress-tests

Commit:
Refactor isValidUsefulLinks().

Commit:
Merge branch 'latest' into 4496-update-useful-links-cypress-tests

Commit:
Merge branch 'latest' into 4496-update-useful-links-cypress-tests

DenisHdz commented on cypress/integration/pages/frontPage/tests.js:
```diff
@@ -168,42 +185,19 @@ export const testsThatFollowSmokeTestConfig = ({ service, pageType }) =>
+              if (isValidUsefulLinks(pageData)) {
+                cy.get('[data-e2e="useful-links"]')
+                  .eq(0)
+                  .within(() => {
```
LGTM, but this could be improved checking if they link is within a `ul li` when there is more than one item, and within a `div` when there is just one.

Bopchy replied:
I thought of that, but was of the opinion that this is checked in the unit tests https://github.com/bbc/simorgh/blob/c35b8d74327bc7cca723f58b33080e66ec6a2d7b/src/app/containers/FrontPageSection/UsefulLinks/index.test.jsx#L7. So the integration test should check if the Useful links appear when they are supposed to, and do not appear when they shouldn't

DenisHdz committed:
Merge branch 'latest' into 4496-update-useful-links-cypress-tests

EinsteinNjoroge committed:
Merge branch 'latest' into 4496-update-useful-links-cypress-tests

Commit:
Add contentType == Guide check.

Commit:
Merge branch 'latest' of github.com:bbc/simorgh into 4496-update-useful-links-cypress-tests

Commit:
Add usefulLinkItems length check.

Commit:
Add usefulLinks test data.

Commit:
Merge branch '4496-update-useful-links-cypress-tests' of github.com:bbc/simorgh into 4496-update-useful-links-cypress-tests

Commit:
Add renderUsefulLinks().

Commit:
Merge branch 'latest' of github.com:bbc/simorgh into 4496-update-useful-links-cypress-tests

Commit:
Add Guide check to filterContentType()

Commit:
Merge branch 'latest' of github.com:bbc/simorgh into 4496-update-useful-links-cypress-tests

Commit:
Remove new line.

jamesbrumpton committed:
Merge branch 'latest' into 4496-update-useful-links-cypress-tests

jamesbrumpton commented:
lgtm

amywalkerdev committed:
Merge branch 'latest' into 4496-update-useful-links-cypress-tests

amywalkerdev committed:
Merge branch 'latest' into 4496-update-useful-links-cypress-tests

amywalkerdev committed:
Merge branch 'latest' into 4496-update-useful-links-cypress-tests
